components/feature/search/src/test/java/mozilla/components/feature/search/ext/BrowserStoreKtTest.kt

```diff
@@ -0,0 +1,101 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.search.ext
+
+import mozilla.components.browser.state.action.SearchAction
+import mozilla.components.browser.state.search.SearchEngine
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.SearchState
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.support.test.mock
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNotNull
+import org.junit.Assert.assertNull
+import org.junit.Assert.assertTrue
+import org.junit.Test
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+
+class BrowserStoreKtTest {
+    @Test
+    fun `waitForDefaultSearchEngine - with state already loaded`() {
+        val store = BrowserStore(
+            BrowserState(
+                search = SearchState(
+                    regionSearchEngines = listOf(
+                        SearchEngine(
+                            id = "google",
+                            name = "Google",
+                            icon = mock(),
+                            type = SearchEngine.Type.BUNDLED
+                        )
+                    ),
+                    defaultSearchEngineId = "google",
+                    complete = true
+                )
+            )
+        )
+
+        val latch = CountDownLatch(1)
+
+        store.waitForDefaultSearchEngine { searchEngine ->
+            assertNotNull(searchEngine)
+            assertEquals("google", searchEngine!!.id)
+            latch.countDown()
+        }
+
+        assertTrue(latch.await(1000, TimeUnit.SECONDS))
+    }
+
+    @Test
+    fun `waitForDefaultSearchEngine - with state dispatched later`() {
+        val store = BrowserStore()
+
+        val latch = CountDownLatch(1)
+
+        store.waitForDefaultSearchEngine { searchEngine ->
+            assertNotNull(searchEngine)
+            assertEquals("google", searchEngine!!.id)
+            latch.countDown()
+        }
+
+        store.dispatch(SearchAction.SetSearchEnginesAction(
+            regionSearchEngines = listOf(
+                SearchEngine(
+                    id = "google",
+                    name = "Google",
+                    icon = mock(),
+                    type = SearchEngine.Type.BUNDLED
+                )
+            ),
+            defaultSearchEngineId = null,
+            regionDefaultSearchEngineId = "google",
+            customSearchEngines = emptyList()
+        ))
```

Comment:
I think we need a `joinBlocking` here? We're giving it a second but still this could fail intermittently, unless I misread?

Reply:
I first thought so too. But the `await()` on the `CountDownLatch` will already do the wait here.

Reply:
@pocmo right but only for 1 second...umm no I misread it's 1000 seconds :).

Reply:
So, what I thought is that `await` on `CountDownLatch` could time out before the action is processed (on a slow build), but if we're waiting long enough that should be good too. I was just concerned that one second might not be enough, but I misread.

Reply:
You read my intention, I definitely wanted to say 1 second here. 😅 I could bump it up, but I would expect the dispatch to have happened in 1 second?

Reply:
Maybe something like 10 seconds to be super conservative? It didn't fail once with 1 seconds in 10k runs locally. My primary goal was to avoid this running forever if this regresses.

Reply:
Alright, I have set it to 10 seconds. :)